### PR TITLE
Additional adversary features

### PR DESCRIPTION
**Pull request by ChaitanyaKonda**
This PR adds the following functionality 
- Adversary can now create a bad deposit as well as a bad transfer transaction with `duplicate commitment`. Earlier the transaction type was randomly picked during adversary test which always ended up being transfer.
- Adversary can now create a bad transfer as well as a bad withdraw transaction with `duplicate nullifier`. Earlier the transaction type was randomly picked during adversary test which always ended up being transfer.
- Adversary can now create a bad deposit, bad transfer and bad withdraw with `incorrect proof`. Earlier the transaction type was randomly picked during adversary test which always ended up being transfer.
- Adversary can now create a bad deposit, bad transfer and bad withdraw with `incorrect public input`. Earlier this wasn’t part of adversary test
- Fixed incorrect `challengeProofVerification` calldata while creating a challenge. As part of `challengeProofVerification` one needs to also pass the L2 block and L2 block transactions data of the historic root. But the L2 block transactions of the historic root were incorrectly passed by sending L2 block transactions of the block being challenged.
- makes optimist container to  be challenger instead of adversary container
-  Fixed few bugs from master
    - add missing mongodb container for adversary optimist container
    - stake passed was undefined for re-registration in adverary.test.mjs
    - logic to store, get transaction siblingPath from database was not working
-  Implemented `signalRollbackCompleted` for challenger as well
-  Fixed state-sync to not to start challenges if optimist is not intent to have a challenger
- A work-around which revert code in state contract proposeBlock assembly code to make adversary test (temporary fix) 

To do

1. Additional testing revealed that deposit transactions that are invalid are not removed at rollback stage and continuously get proposed as they are set to be in `mempool` 
2. When multiple L2 blocks are assembled at once and if they are all wrong blocks, then all of these block proposed events are handled one after another and challenges get created for all of them, even though challenging just the earliest is sufficient. As a result, each of these wrong bad block proposed events will add a stop event in stop queue. Until, all of these are removed from the stop queue by a rollback event, proposer cannot propose blocks. Also, the subsequent challenges after the earlier challenge will fail and not produce a rollback because that bad L2 block does not exist anymore. This needs to be handled by such that any block proposed events on top of bad block are not processed until the first rollback event. This needs to be ensured for initial state sync and during resend too
3. IncorrectHistoricRoot should have functionality to pass incorrect root for transfer and withdraw transactionTypes which will be picked by the tester and not random

Testing:
Each of the bad transaction type when tested individually passes. The above issues when fixed will enable the bad blocks to all be successfully tested and subsequently too. For the moment, just one bad transaction type `DuplicateNullifierWithdraw` is turned on

### CODE REVIEW

**Comment by ChaitanyaKonda** [comment]
DNM due to the to do list specified here

File: nightfall-optimist/src/services/challenges.mjs [hunk 1]
```diff
@@ -108,7 +133,7 @@ export async function createChallenge(block, transactions, err) {
     }
     // Challenge wrong root
     case 1: {
-      logger.debug('Challenging incorrect root');
+      logger.debug(`Challenging incorrect leaf count for block ${JSON.stringify(block, null, 2)}`);
```
Comment (RogerTaule) [inline nightfall-optimist/src/services/challenges.mjs:136]: This logger should be `Challenging incorrect root`

**Review by RogerTaule — CHANGES_REQUESTED** [review]
It looks good to me. 
I have added a couple of comments to remove some console.logs, but overall it is good. 
Would it be possible to squash the commits so that only one commit is merged? I think it will make it easier for future rebases.

**Comment by Westlad** [comment]
Good for me.  I'll approve once @RogerTaule's comments are cleared.

File: nightfall-optimist/src/event-handlers/block-proposed.mjs [hunk 2]
```diff
@@ -116,7 +116,16 @@ async function blockProposedEventHandler(data) {
...
-    await checkBlock(block, transactions);
+    // we should not check the block if the stop queue is not empty because
+    // it signals that there is a bad block which will get challenged eventually
+    // meanwhile any new L2 blocks received if turned out to be bad blocks will
+    // raise a commit to challenge and reveal challenge which is bound to fail because
+    // a rollback from previous wrong block would have removed this anyway.
+    // Instead, what happens now is that any good/bad blocks on top of the first bad block
+    // will get saved and eventually all these blocks will be removed as part of the rollback
+    // of the first bad block
+    console.log('HERE queues[2].length', queues[2].length);
```
[the remark this answers is not in the document]
Comment (LijuJoseJJ) [reply nightfall-optimist/src/event-handlers/block-proposed.mjs]: removed

File: nightfall-optimist/src/event-handlers/block-proposed.mjs [hunk 3]
```diff
@@ -143,6 +152,7 @@ async function blockProposedEventHandler(data) {
       // have the actual challenge to support syncing
       logger.debug('enqueuing event to stop queue');
       await enqueueEvent(commitToChallenge, 2, txDataToSign);
+      console.log('HERE queues[2].length in catch', queues[2].length);
```
[the remark this answers is not in the document]
Comment (LijuJoseJJ) [reply nightfall-optimist/src/event-handlers/block-proposed.mjs]: removed

File: nightfall-optimist/src/services/challenges.mjs (hunk 1 above)
Comment (LijuJoseJJ) [reply nightfall-optimist/src/services/challenges.mjs]: changed

**Review by RogerTaule — APPROVED** [review]

**Review by Westlad — APPROVED** [review]

**Pull request merged**